pkgs/native_assets_cli/lib/src/config.dart

```diff
@@ -404,6 +406,166 @@ class BuildOutputBuilder extends HookOutputBuilder {
+  ///
+  /// If any specified path does not exist, a [FileSystemException] is thrown.
+  /// Any error during the directory listing is caught and rethrown with
+  /// additional context.
+  ///
+  /// When recursive is set to true, the method will also add all subdirectories
+  /// and their files as dependencies.
+  Future<void> addDataAssetDirectories(
+    List<String> paths, {
+    required BuildInput input,
+    bool recursive = false,
+  }) async {
+    final packageName = input.packageName;
+    final packageRoot = input.packageRoot;
+    for (final path in paths) {
+      final resolvedUri = packageRoot.resolve('$packageName/$path');
+      final directory = Directory.fromUri(resolvedUri);
+      final file = File.fromUri(resolvedUri);
+
+      if (await directory.exists()) {
+        // Process as a directory.
+        addDependency(directory.uri);
+        try {
+          await for (final entity in directory.list(
+            recursive: recursive,
+            followLinks: false,
+          )) {
+            // Add dependency for every file and directory found.
+            addDependency(entity.uri);
+          }
+        } on FileSystemException catch (e) {
+          throw FileSystemException(
+            'Error reading directory "$path": ${e.message}',
+            directory.path,
+            e.osError,
+          );
+        }
+      } else if (await file.exists()) {
+        addDependency(file.parent.uri);
+        addDependency(file.uri);
+      } else {
+        throw FileSystemException(
+          'Path does not exist',
+          resolvedUri.toFilePath(windows: Platform.isWindows),
+        );
+      }
+    }
+  }
+}
+
+extension AddFoundCodeAssetsExtension on BuildOutputBuilder {
+  /// Searches recursively through the entire expected output directory
+  /// for native library files that match the expected target filename
+  /// based on a list of asset mappings. An output directory can be optionally
+  /// provided, otherwise the output directory from the input is used.
+  ///
+  /// Each mapping in [assetMappings] is a Map with a single key-value pair
+  /// where:
+  /// - The key represents the filename (given without prefix and extension) to
+  ///   search for.
+  /// - The value represents the linked CodeAsset name that should be used.
+  ///
+  /// The expected filename is computed using the current operating system's
+  /// naming conventions and a concrete [LinkMode] derived from the input's
+  /// link preference.
+  ///
+  /// For each file that ends with the computed library filename for one of the
+  /// provided mappings, a [CodeAsset] is created and added via
+  /// [addEncodedAsset] if it hasn't already been added.
+  ///
+  /// Returns a list of URIs corresponding to all the added code assets.
+  Future<List<Uri>> addFoundCodeAssets({
+    required BuildInput input,
+    required List<Map<String, String>> assetMappings,
+    Uri? outputDirectory,
```

Comment:
I'd make this `required Uri directory`.

That way we can also point it to directories that are not output directories. (For example if someone were to decide to check in generated binaries in their project in git 🙈  or if someone would use a package manager to download binaries and you want to pass the directory of the package manager.)

Reply:
This was nullable to have a default to the input's out directory.

final outDir = outputDirectory ?? input.outputDirectory;